components/sup/src/manager/service.rs

```diff
@@ -810,25 +814,36 @@ impl Service {
         }
     }
 
-    fn post_run(&mut self) {
-        let _timer = hook_timer("post-run");
-
+    fn post_run(&mut self, executor: &TaskExecutor) {
         if let Some(ref hook) = self.hooks.post_run {
-            hook.run(&self.service_group,
-                     &self.pkg,
-                     self.svc_encrypted_password.as_ref());
+            let hook_runner = HookRunner::new(Arc::clone(&hook),
+                                              self.service_group.clone(),
+                                              self.pkg.clone(),
+                                              self.svc_encrypted_password.clone());
+            let (handle, f) = hook_runner.loop_future();
+            self.post_run_handle = Some(handle);
+            executor.spawn(f);
+        }
+    }
+
+    /// Stop the `post-run` future. The `post-run` hook has retry logic based on its exit code. This
+    /// will stop this retry loop regardless of `post-run`'s exit code.
+    ///
+    /// Consumes the handle to that future in the process.
+    fn stop_post_run(&mut self) {
+        if let Some(h) = self.post_run_handle.take() {
+            h.terminate();
         }
     }
 
     // This hook method looks different from all the others because
     // it's the only one that runs async right now.
```

Comment:
We can probably remove this particular comment, since it's no longer the only async hook :joy: